Detectors/EMCAL/simulation/src/Digitizer.cxx

```diff
@@ -0,0 +1,128 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "EMCALSimulation/Digitizer.h"
+#include "EMCALBase/Digit.h"
+#include "EMCALBase/Geometry.h"
+#include "EMCALBase/GeometryBase.h"
+#include "EMCALBase/Hit.h"
+#include "MathUtils/Cartesian3D.h"
+#include "SimulationDataFormat/MCCompLabel.h"
+
+#include <TRandom.h>
+#include <TVector3.h>
+#include <climits>
+#include <forward_list>
+#include "FairLogger.h" // for LOG
+
+ClassImp(o2::EMCAL::Digitizer);
+
+// using o2::EMCAL::GeometryBase;
+using o2::EMCAL::Digit;
+using o2::EMCAL::Hit;
+
+using namespace o2::EMCAL;
+
+//_______________________________________________________________________
+void Digitizer::init() {}
+
+//_______________________________________________________________________
+void Digitizer::finish() {}
+
+//_______________________________________________________________________
+void Digitizer::process(const std::vector<Hit>* hits, std::vector<Digit>* digits)
+{
+  digits->clear();
+  mDigits.clear();
+
+  for (auto hit : *hits) {
+    Digit digit = HitToDigit(hit);
+    Int_t id = digit.GetTower();
+
+    if (id < 0 || id > mGeometry->GetNCells()) {
+      LOG(WARNING) << "tower index out of range: " << id << FairLogger::endl;
+      continue;
+    }
+
+    Bool_t flag = false;
+    for (auto rit = mDigits[id].rbegin(); rit != mDigits[id].rend(); ++rit) {
```

Comment:
Do you need reverse iteration? In any case, one might use a range-based loop also here (avoiding iterators):
```
 for (auto& digit  : boost::adaptors::reverse(mDigits[id])) {
// ...
 }
```

Reply:
To be fixed.

Reply:
The issue here is that for each tower (element of mDigits), the digits are ordered in time with the most recent digit at the end of the vector. (This is how it should be in normal operation, but doing the loop allows for the possibility that the digits are not time-ordered.) Because std::vector does not have a push_front function, I need to use push_back, which means that I need to iterate over the loop in reverse if I want to see the most recent digit first. It seems that there would be three options:

1.) Use the adaptors as Sandro suggested.
2.) Change mDigits from std::unordered_map<Int_t, std::vector<Digit>> to std::unordered_map<Int_t, std::deque<Digit>> or std::unordered_map<Int_t, std::list<Digit>> . Then I could use push_front to add the most recent digit and iterate forwards over the digits.
3.) Assume that the digits are time-ordered and only check the last digit in the vector. This would also be faster (no loop), but would cause problems if the Hits/Digits were not time-ordered for some reason.

Please let me know what you think. If adaptors are preferred, I will need some help implementing them.

Reply:
If a content of a container should be ordered I would propose to have an ordered container, therefore voting for option2. Eventually std::set with the proper comparator is the better option, although this probably needs the operator> to be implemented as well (which is not a big deal).

Reply:
I would avoid `std::set` has it has a big per item overhead and it's sparse in memory, which kills cache efficiency. IMHO it's much better to use and std::vector and sort it as required.

Reply:
What about deque or list rather than set?

Reply:
A list is also sparse in memory and bad for caching. deque might be a compromise. In any case, I would avoid `std::set`.